monai/losses/dice.py

```diff
@@ -298,5 +300,177 @@ def forward(self, input: torch.Tensor, target: torch.Tensor, smooth: float = 1e-
         return f
 
 
+class GeneralizedWassersteinDiceLoss(_Loss):
+    """
+    Generalized Wasserstein Dice Loss [1] in PyTorch.
+    Compared to [1] we used a weighting method similar to the one
+    used in the generalized Dice Loss [2].
+
+    References:
+    ===========
+    [1] "Generalised Wasserstein Dice Score for Imbalanced Multi-class
+        Segmentation using Holistic Convolutional Networks",
+        Fidon L. et al. MICCAI BrainLes 2017.
+    [2] "Generalised dice overlap as a deep learning loss function
+        for highly unbalanced segmentations",
+        Sudre C., et al. MICCAI DLMIA 2017.
+
+    wasserstein_distance_map:
+    Compute the voxel-wise Wasserstein distance (eq. 6 in [1]) between the
+    flattened prediction and the flattened labels (ground_truth) with respect
+    to the distance matrix on the label space M.
+    References:
+    [1] "Generalised Wasserstein Dice Score for Imbalanced Multi-class
+        Segmentation using Holistic Convolutional Networks",
+        Fidon L. et al. MICCAI BrainLes 2017
+
+    compute_weights_generalized_true_positives:
+    Compute the weights \alpha_l of eq. 9 in [1] but using the weighting
+    method proposed in the generalized Dice Loss [2].
+    References:
+    [1] "Generalised Wasserstein Dice Score for Imbalanced Multi-class
+        Segmentation using Holistic Convolutional Networks",
+        Fidon L. et al. MICCAI BrainLes 2017
+    [2] "Generalised dice overlap as a deep learning loss function
+        for highly unbalanced segmentations." Sudre C., et al.
+        MICCAI DLMIA 2017.
+    """
+
+    def __init__(self, dist_matrix, reduction="mean"):
+        """
+        Args:
+            param dist_matrix: 2d tensor or 2d numpy array; matrix of distances
+            between the classes. It must have dimension C x C where C is the
+            number of classes.
+            param reduction: str; reduction mode.
+
+        Raises:
+            ValueError: dist_matrix.shape[0] != dist_matrix.shape[1] is invalid.
+
+        """
+        super(GeneralizedWassersteinDiceLoss, self).__init__(reduction=reduction)
+
+        if dist_matrix.shape[0] != dist_matrix.shape[1]:
+            raise ValueError("Dist Matrix is invalid.")
+
+        self.M = dist_matrix
+        if isinstance(self.M, np.ndarray):
+            self.M = torch.from_numpy(self.M)
+        if torch.max(self.M) != 1:
+            self.M = self.M / torch.max(self.M)
+        self.num_classes = self.M.size(0)
+        self.reduction = reduction
```

Comment:
`self.reduction = reduction` is done in the parent class already, I think we can remove line 362

Reply:
I am not using DiceLoss as the parent class here. I am using _Loss from torch.

Reply:
_Loss also has a reduction variable, could follow this https://github.com/Project-MONAI/MONAI/blob/67741bfd8b6809be299658acd787f01f520cbd6a/monai/losses/dice.py#L69

Reply:
@wyli thanks will look into this. also how do I properly cite the resource I am referring to? Like the tests are failing because I am referring to the citation from where I referred the code.

# @inproceedings{fidon2017generalised,
#  title={Generalised {W}asserstein dice score for imbalanced multi-class 
#         segmentation using holistic convolutional networks},
#  author={Fidon, Lucas and Li, Wenqi and Garcia-Peraza-Herrera, Luis C 
#          and Ekanayake, Jinendra and Kitchen, Neil and Ourselin, 
#          S{\'e}bastien and #Vercauteren, Tom},
#  booktitle={International MICCAI Brainlesion Workshop},
#  pages={64--76},
#  year={2017},
#  organization={Springer}
# }

Reply:
@wyli thanks will look into this. also how do I properly cite the resource I am referring to? Like the tests are failing because I am referring to the citation from where I referred the code.

"""
@inproceedings{fidon2017generalised,
title={Generalised {W}asserstein dice score for imbalanced multi-class segmentation using holistic convolutional networks},
author={Fidon, Lucas and Li, Wenqi and Garcia-Peraza-Herrera, Luis C and Ekanayake, Jinendra and Kitchen, Neil and Ourselin, S{\'e}bastien and #Vercauteren, Tom},
booktitle={International MICCAI Brainlesion Workshop},pages={64--76}, year={2017},
organization={Springer}
"""

Reply:
those test cases are not from the publication, so you can just remove the paragraph. the `GeneralizedWassersteinDiceLoss` you've created already nicely covers all the reference information

Reply:
ok thanks. working on it.